src/controllers/wallet/atomic-swap/tx-proposal.controller.js

```diff
@@ -0,0 +1,292 @@
+  }
+}
+
+async function signAndPush(req, res) {
+  const validationResult = parametersValidation(req);
+  if (!validationResult.success) {
+    res.status(400).json(validationResult);
+    return;
+  }
+
+  const canStart = lock.lock(lockTypes.SEND_TX);
+  if (!canStart) {
+    res.send({ success: false, error: cantSendTxErrorMessage });
+    return;
+  }
+
+  const network = req.wallet.getNetworkObject();
+
+  const partialTx = req.body.partial_tx;
+  const signatures = req.body.signatures || [];
+
+  const proposal = PartialTxProposal.fromPartialTx(partialTx, network);
+
+  try {
+    storage.setStore(req.wallet.store);
+    await proposal.signData('123');
+    for (const signature of signatures) {
+      proposal.signatures.addSignatures(signature);
+    }
+    if (!proposal.isComplete()) {
+      throw new Error('Transaction is not complete');
+    }
+    const tx = proposal.prepareTx();
+
+    const sendTransaction = new SendTransaction({ transaction: tx, network });
+    const response = await sendTransaction.runFromMining();
+    res.send({ success: true, ...mapTxReturn(response) });
+  } catch (err) {
+    res.send({ success: false, error: err.message });
+  } finally {
+    lock.unlock(lockTypes.SEND_TX);
+  }
+}
+
+async function getInputData(req, res) {
+  const validationResult = parametersValidation(req);
+  if (!validationResult.success) {
+    res.status(400).json(validationResult);
+    return;
+  }
+
+  const network = req.wallet.getNetworkObject();
+  const { txHex } = req.body;
+
+  try {
+    const tx = helpersUtils.createTxFromHex(txHex, network);
+    const signatures = new PartialTxInputData(tx.getDataToSign().toString('hex'), tx.inputs.length);
+
+    for (const [index, input] of tx.inputs.entries()) {
+      if ((!input.data) || input.data.length === 0) {
+        // input without data to be added
+        continue;
+      }
+
+      signatures.addData(index, input.data);
+    }
+
+    res.send({ success: true, signatures: signatures.serialize() });
+  } catch (err) {
+    res.send({ success: false, error: err.message });
+  }
+}
+
+async function getLockedUTXOs(req, res) {
+  try {
+    const utxos = [];
+    const historyTransactions = req.wallet.getFullHistory();
+    for (const tx of Object.values(historyTransactions)) {
+      const marked = [];
+      for (const [index, output] of tx.outputs.entries()) {
+        if (output.selected_as_input === true) {
+          marked.push(index);
+        }
+      }
+      if (marked.length !== 0) {
+        utxos.push({
+          tx_id: tx.tx_id,
+          outputs: marked,
+        });
+      }
+    }
+
+    res.send({ success: true, locked_utxos: utxos });
+  } catch (err) {
+    res.send({ success: false, error: err.message });
+  }
+}
+
+async function unlockInputs(req, res) {
+  const validationResult = parametersValidation(req);
+  if (!validationResult.success) {
+    res.status(400).json(validationResult);
+    return;
+  }
+
```

Comment:
I would create the same `SEND_TX` lock here because this `selected_as_input` parameter might be affected if a tx is being sent in the moment

Reply:
Done.